File: runtimes/java/src/main/java/wrapper/ValidateHooks.java
package wrapper;

import com.google.gson.Gson;
import java.net.http.HttpClient;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * T041: Java runtime validate-support hooks (FR-021, FR-022). Mirrors the Node/Python hooks:
 * invoke every in-scope tool once, trigger the OAuth flow once, and confirm retry/logging fires.
 * Never touches an out-of-scope tool.
 */
public final class ValidateHooks {
  private static final Gson GSON = new Gson();

  private ValidateHooks() {}

  public static final class ToolInvocationOutcome {
    public final String toolName;
    public final boolean invoked;
    public final boolean success;
    public final String error;

    public ToolInvocationOutcome(String toolName, boolean invoked, boolean success, String error) {
      this.toolName = toolName;
      this.invoked = invoked;
      this.success = success;
      this.error = error;
    }
  }

  private static Map<String, Object> sampleArgs(Manifest.ToolDefinition tool) {
    Map<String, Object> args = new HashMap<>();
    if (tool.inputSchema != null && tool.inputSchema.has("properties")) {
      var props = tool.inputSchema.getAsJsonObject("properties");
      for (String name : props.keySet()) {
        var prop = props.getAsJsonObject(name);
        String type = prop.has("type") ? prop.get("type").getAsString() : "string";
        args.put(name, "number".equals(type) ? 1 : "test");
      }
    }
    return args;
  }

  public static List<ToolInvocationOutcome> invokeAllInScope(
      Manifest.LoadedPackage pkg, String baseUrl, HttpClient client) {
    List<ToolInvocationOutcome> results = new ArrayList<>();
    for (Manifest.ToolDefinition tool : Manifest.inScopeTools(pkg.tools, pkg.scope)) {
      Reliability.InvokeResult res =
          Server.dispatchTool(tool, sampleArgs(tool), baseUrl, client);
      results.add(new ToolInvocationOutcome(tool.name, true, res.ok, res.ok ? null : res.error));
    }
    return results;
  }

  public static boolean exerciseOAuthOnce(Manifest.LoadedPackage pkg) {
    var c = pkg.oauthConfig;
    return c.has("authorizationEndpoint") && c.has("tokenEndpoint") && c.has("redirectMode");
  }

  public static boolean verifyRetryBehavior() {
    return Reliability.isRetryable(true, 503, false) && Reliability.backoffSchedule().size() == 2;
  }
}

File: runtimes/python/src/wrapper_runtime/validate_hooks.py
"""T040: Python runtime validate-support hooks (FR-021, FR-022).

Mirrors runtimes/node/src/validate-hooks.ts: invoke every in-scope tool once, trigger
the OAuth flow once, and confirm retry/logging fires. Never touches an out-of-scope tool.
"""
from __future__ import annotations

import json
import urllib.request
from typing import Any

from .manifest import LoadedPackage, ToolDefinition, in_scope_tools
from .reliability import backoff_schedule, is_retryable
from .server import invoke_with_reliability, tool_to_request


def _sample_args(tool: ToolDefinition) -> dict[str, Any]:
    args: dict[str, Any] = {}
    for name, prop in (tool.inputSchema.get("properties") or {}).items():
        args[name] = 1 if prop.get("type") == "number" else "test"
    return args


def invoke_all_in_scope(pkg: LoadedPackage, base_url: str) -> list[dict[str, Any]]:
    """Invoke every in-scope tool once against the wrapped webapp."""
    results: list[dict[str, Any]] = []
    for tool in in_scope_tools(pkg.tools, pkg.scope):
        url, method = tool_to_request(tool, _sample_args(tool), base_url)

        def call(_attempt: int, _url: str = url, _method: str = method) -> tuple[int, Any]:
            req = urllib.request.Request(_url, method=_method)
            with urllib.request.urlopen(req) as resp:  # noqa: S310
                raw = resp.read().decode("utf-8")
                try:
                    return resp.status, json.loads(raw)
                except json.JSONDecodeError:
                    return resp.status, raw

        res = invoke_with_reliability(tool.name, tool.annotations["idempotentHint"], call)
        entry = {"toolName": tool.name, "invoked": True, "success": res["ok"]}
        if not res["ok"] and res.get("error"):
            entry["error"] = res["error"]
        results.append(entry)
    return results


def exercise_oauth_once(pkg: LoadedPackage) -> bool:
    c = pkg.oauth_config
    return bool(c.get("authorizationEndpoint") and c.get("tokenEndpoint") and c.get("redirectMode"))


def verify_retry_behavior() -> bool:
    return is_retryable(idempotent=True, status=503) and len(backoff_schedule()) == 2
